spec/lib/signon/bootstrap_spec.rb

```diff
@@ -32,80 +10,228 @@
-             )
-
-      allow_any_instance_of(Aws::SecretsManager::Client).to \
-        receive(:put_secret_value)
-        .with(
-          secret_id: arn,
-          secret_string: generated_secret,
-          version_stages: %w[AWSCURRENT],
-        )
-
-      expect { call }.not_to raise_error
-      expect(stub).to have_been_requested
+  describe "#bootstrap_applications" do
+    subject(:call) do
+      logger = Logger.new($stdout)
+      logger.level = Logger::WARN
+      described_class.bootstrap_applications(
+        applications: [
+          {
+            "name" => name,
+            "id_arn" => id_arn,
+            "secret_arn" => secret_arn,
+            "permissions" => permissions,
+            "description" => description,
+            "home_uri" => home_uri,
+            "redirect_uri" => redirect_uri,
+          },
+        ],
+        signon: signon_client,
+        secretsmanager: secretsmanager_client,
+        logger: logger,
+      )
     end
-  end
 
-  def stub_get_secret
-    allow_any_instance_of(Aws::SecretsManager::Client).to \
-      receive(:get_secret_value)
+    let(:name) { "Publishing API" }
+    let(:id_arn) { "arn:id:123" }
+    let(:secret_arn) { "arn:secret:123" }
+    let(:permissions) { %w[signin] }
+    let(:description) { "Content Management System backend" }
+    let(:home_uri) { "https://pub-api.example.org" }
+    let(:redirect_uri) { "https://pub-api.example.org/redirect" }
+    let(:oauth_id) { "oauth_id" }
+    let(:oauth_secret) { "oauth_secret" }
+
+    context "when the application has not been created" do
+      let(:versions) do
+        {}
+      end
+
+      it "will create the application and put the secret in SecretsManager" do
+        expect(secretsmanager_client).to receive(:describe_secret)
+          .with(secret_id: id_arn)
+          .and_return(metadata)
+
+        expect(signon_client).to receive(:create_application)
+          .with(
+            name: name,
+            description: description,
+            home_uri: home_uri,
+            permissions: permissions,
+            redirect_uri: redirect_uri,
+          )
+          .and_return({ "oauth_id" => oauth_id, "oauth_secret" => oauth_secret })
+
+        expect(secretsmanager_client).to receive(:put_secret_value)
+          .with(
+            secret_id: secret_arn,
+            secret_string: oauth_secret,
+            version_stages: %w[AWSCURRENT],
+          )
+
+        expect(secretsmanager_client).to receive(:put_secret_value)
+          .with(
+            secret_id: id_arn,
+            secret_string: oauth_id,
+            version_stages: %w[AWSCURRENT],
+          )
+
+        expect { call }.not_to raise_error
+      end
+    end
+
+    context "when application has been created" do
+      context "when credentials are in secretsmanager" do
+        let(:versions) do
+          { "version1" => %w[AWSCURRENT] }
+        end
+
+        it "will verify the credentials have been set in SecretsManager" do
```

Comment:
"and not create a new application or secret"? (are there length limits on these things?)